### PR TITLE
iq-tree method

### PR DESCRIPTION
**Pull request by mikerobeson**
I've made two new phylogenetic tree constructions methods for `q2-phylogeny` based on [IQ-TREE](http://www.iqtree.org/) v1.5.5 as available from [bioconda](https://anaconda.org/bioconda/iqtree).

The two methods are:
1) `iqtree`: returns a single tree.
2) `iqtree-ultrafast-bootstrap`: similar to the existing `raxml-rapid-bootstrap`. This will return a tree with bootstrap supports.

### CODE REVIEW

File: ci/recipe/meta.yaml [hunk 1]
```diff
@@ -25,6 +25,7 @@ requirements:
     - q2-types {{ release }}.*
     - fasttree
     - raxml
+    - iqtree
```
Comment (thermokarst) [inline ci/recipe/meta.yaml]: Not sure why this is showing up in the diff still --- I merged #47 ...

File: q2_phylogeny/_iqtree.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+import tempfile
+
+from random import randint
+
+from q2_types.feature_data import AlignedDNAFASTAFormat
+from q2_types.tree import NewickFormat
+from q2_phylogeny._raxml import run_command
+
+
+def _build_iqtree_command(alignment, seed,
```
[the remark this answers is not in the document]
Comment (mikerobeson) [reply q2_phylogeny/_iqtree.py]: Done. I hope I emulated the example properly!

File: q2_phylogeny/_iqtree.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+                          stop_iter=None,
+                          perturb_nni_strength=None,
+                          spr_radius=None,
+                          allnni=False,
+                          safe=False):
+
+    cmd = ['iqtree', '-nt', '%i' % n_threads]
+
+    if seed is None:
+        seed = randint(1000, 10000)
```
[the remark this answers is not in the document]
Comment (mikerobeson) [reply q2_phylogeny/_iqtree.py]: Agreed. I set the option to be ignored if not supplied.

File: q2_phylogeny/_iqtree.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    if min_cor_ufboot:
+        cmd += ['-bcor', '%f' % min_cor_ufboot]
+
+    if ep_break_ufboot:
+        cmd += ['-beps', '%f' % ep_break_ufboot]
+
+    return cmd
+
+
+def iqtree_ultrafast_bootstrap(alignment: AlignedDNAFASTAFormat,
```
[the remark this answers is not in the document]
Comment (mikerobeson) [reply q2_phylogeny/_iqtree.py]: Done. I hope I emulated the example properly!

File: q2_phylogeny/_iqtree.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+                               min_cor_ufboot=None,
+                               ep_break_ufboot=None,
+                               allnni=False,
+                               safe=False):
+    # This is a separate command becuase there are several
+    # bootstrap specific options.
+    cmd = ['iqtree', '-nt', '%i' % n_threads, '-bb',
+           '%i' % bootstrap_replicates]
+
+    if seed is None:
```
[the remark this answers is not in the document]
Comment (mikerobeson) [reply q2_phylogeny/_iqtree.py]: Agreed. I set the `-seed` option to be ignored if not supplied.

File: q2_phylogeny/plugin_setup.py [hunk 6]
```diff
@@ -130,6 +188,142 @@
...
+                                 'If not set, program defaults will be used. '
+                                 'See IQ-TREE manual for details.'),
+        'spr_radius': ('Radius for parsimony SPR search. If not set, '
+                       'program defaults will be used. See IQ-TREE manual '
+                       'for details.'),
+        'allnni': ('Perform more thorough NNI search.'),
+        'safe': ('Safe likelihood kernel to avoid numerical underflow')},
+    output_descriptions={'tree': 'The resulting phylogenetic tree.'},
+    name='Construct a phylogenetic tree with IQ-TREE.',
+    description=('Construct a phylogenetic tree with IQ-TREE with automatic '
```
[the remark this answers is not in the document]
Comment (mikerobeson) [reply q2_phylogeny/plugin_setup.py]: Done. Added link http://www.iqtree.org/.

File: q2_phylogeny/_iqtree.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+                               spr_radius=None,
+                               n_max_ufboot_iter=None,
+                               n_ufboot_steps=None,
+                               min_cor_ufboot=None,
+                               ep_break_ufboot=None,
+                               allnni=False,
+                               safe=False):
+    # This is a separate command becuase there are several
+    # bootstrap specific options.
+    cmd = ['iqtree', '-nt', '%i' % n_threads, '-bb',
```
[the remark this answers is not in the document]
Comment (mikerobeson) [reply q2_phylogeny/_iqtree.py]: Great idea! I have now set `0` == `AUTO`.

File: ci/recipe/meta.yaml (hunk 1 above)
Comment (mikerobeson) [reply ci/recipe/meta.yaml]: My fault. I forgot to sync the repo.

File: q2_phylogeny/plugin_setup.py [hunk 8]
```diff
@@ -211,12 +211,13 @@
     },
     parameter_descriptions={
         'n_threads': ('The number of threads to use for multithreaded '
-                      'processing.'),
-        'substitution_model': ('Model of Nucleotide Substitution.'
+                      'processing. Use \'0\' to let IQ-TREE automatically '
+                      'determine the optimal numbr of cores to use.'),
```
Comment (thermokarst) [inline q2_phylogeny/plugin_setup.py]: Minor typo  "numbr"

File: q2_phylogeny/plugin_setup.py [hunk 9]
```diff
@@ -193,7 +193,7 @@
     inputs={'alignment': FeatureData[AlignedSequence]},
     parameters={
             'seed': Int,
-            'n_threads': Int % Range(1, None),
+            'n_threads': Int,
```
Comment (thermokarst) [inline q2_phylogeny/plugin_setup.py]: Ah, missed this early, the iqtree docs indicate that this is multicore processing, not multi-threaded. We should probably call this parameter `n_cores`. The support text could say:

> The number of CPU cores to use for processing. Use '0' to let IQ-TREE automatically determine the optimal number of cores to use.

File: q2_phylogeny/plugin_setup.py [hunk 10]
```diff
@@ -252,7 +253,7 @@
...
```
Comment (thermokarst) [inline q2_phylogeny/plugin_setup.py]: Same as above. Sorry for missing this!

File: q2_phylogeny/plugin_setup.py [hunk 11]
```diff
@@ -274,12 +275,13 @@
     },
     parameter_descriptions={
         'n_threads': ('The number of threads to use for multithreaded '
-                      'processing.'),
+                      'processing. Use \'0\' to let IQ-TREE automatically '
+                      'determine the optimal numbr of cores to use.'),
```
Comment (thermokarst) [inline q2_phylogeny/plugin_setup.py]: Minor typo "numbr"

File: q2_phylogeny/plugin_setup.py (hunk 9 above)
Comment (mikerobeson) [reply q2_phylogeny/plugin_setup.py]: Right! I fixed this throughout the code / test code.

File: q2_phylogeny/plugin_setup.py (hunk 8 above)
Comment (mikerobeson) [reply q2_phylogeny/plugin_setup.py]: Nice catch!

**Comment by thermokarst** [comment]
Woohoo, thanks @mikerobeson! This looks great! Super excited to have more tree-building options in QIIME 2!

**Comment by thermokarst** [comment]
@mikerobeson - want to edit the placeholder I added to the [changelog](https://forum.qiime2.org/t/preview-qiime-2-2018-8-development-changelog/4724?u=thermokarst)? It should be editable for you, if not please let me know!

**Comment by mikerobeson** [comment]
Thanks @thermokarst! I slightly reworded the text in the changelog. I'll eventually add IQ-TREE related content to the [q2-phylogeny community tutorial](https://forum.qiime2.org/t/q2-phylogeny-community-tutorial/4455).

**Comment by thermokarst** [comment]
Cool! Once you are ready we can start to come up with a plan for migrating that tutorial to the official tutorials, if you want. Just let us know what you want to do!

**Pull request merged**